internal/hcsoci/resources.go

```diff
@@ -33,3 +40,120 @@ func NormalizeMemorySize(ctx context.Context, cid string, requestedSizeMB uint64
+// except it uses container resource annotations.
+func NormalizeContainerUpdateResourcesRequest(ctx context.Context, data interface{}, annots map[string]string) (interface{}, error) {
+	return normalizeUpdateResourcesRequest(ctx, true, data, annots)
+}
+
+func normalizeUpdateResourcesRequest(ctx context.Context, isContainer bool, data interface{}, annots map[string]string) (interface{}, error) {
+	var (
+		memorySizeInMB  = annotations.MemorySizeInMB
+		processorCount  = annotations.ProcessorCount
+		processorLimit  = annotations.ProcessorLimit
+		processorWeight = annotations.ProcessorWeight
+	)
+	if isContainer {
+		memorySizeInMB = annotations.ContainerMemorySizeInMB
+		processorCount = annotations.ContainerProcessorCount
+		processorLimit = annotations.ContainerProcessorLimit
+		processorWeight = annotations.ContainerProcessorWeight
+	}
+
+	// could use [ConvertCPULimits], but that would only work for containers, and wouldn't distinguish between
+	// an annotation being set to "0", and the annotation not being present.
+	// And would still need to write out logic for uVM CPU annotation parsing.
+	//
+	// Don't bother checking if multiple CPU values are set for containers, since hcsTask will check in [updateWCOWResources].
+	switch resources := data.(type) {
+	case *specs.WindowsResources:
+		if x := oci.ParseAnnotationsUint64(ctx, annots, memorySizeInMB, 0); x != 0 {
+			if resources.Memory == nil {
+				resources.Memory = &specs.WindowsMemoryResources{}
+			}
+			if y := resources.Memory.Limit; y != nil && *y > 0 {
+				return nil, fmt.Errorf("conflicting memory limits between resource limit (%d) and %q annotation (%d)", *y, memorySizeInMB, x)
+			}
+			xx := x * memory.MiB
+			resources.Memory.Limit = &xx
+		}
+
+		if x := oci.ParseAnnotationsUint64(ctx, annots, processorCount, 0); x != 0 {
+			if resources.CPU == nil {
+				resources.CPU = &specs.WindowsCPUResources{}
+			}
+
+			if y := resources.CPU.Count; y != nil && *y > 0 {
+				return nil, fmt.Errorf("conflicting CPU count between resource limit (%d) and %q annotation (%d)", *y, processorCount, x)
+			}
+			resources.CPU.Count = &x
+		}
+		if x := oci.ParseAnnotationsUint64(ctx, annots, processorLimit, 0); x != 0 {
+			if resources.CPU == nil {
+				resources.CPU = &specs.WindowsCPUResources{}
+			}
+			if y := resources.CPU.Maximum; y != nil && *y > 0 {
+				return nil, fmt.Errorf("conflicting CPU maximum between resource limit (%d) and %q annotation (%d)", *y, processorLimit, x)
+			}
+			xx := uint16(x)
+			resources.CPU.Maximum = &xx
+		}
+		if x := oci.ParseAnnotationsUint64(ctx, annots, processorWeight, 0); x != 0 {
+			if resources.CPU == nil {
+				resources.CPU = &specs.WindowsCPUResources{}
+			}
+			if y := resources.CPU.Shares; y != nil && *y > 0 {
+				return nil, fmt.Errorf("conflicting CPU shares between resource limit (%d) and %q annotation (%d)", *y, processorWeight, x)
+			}
```

Comment:
For CPU, couldn't y be 0 and still be a valid conflict with the annotation set cpu value?